Use `Action::Stop` in more places

hamishknight opened the pull request:
This used to be something we couldn't express in certain walker methods. Update various places where we can now use it.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/IDE/SwiftSourceDocInfo.cpp:
```diff
@@ -261,28 +263,17 @@ ASTWalker::PreWalkAction NameMatcher::walkToDeclPre(Decl *D) {
 }
 
 ASTWalker::PreWalkResult<Stmt *> NameMatcher::walkToStmtPre(Stmt *S) {
+  if (isDone())
+    return Action::Stop();
```
Are the pre-checks for `isDone` required any more? Only `tryResolve` can empty `LocsToResolve`, so I would have thought that we stop before ever calling another `walk` after it's emptied (from all the stop checks at the *end* of various functions).

hamishknight replied:
I was tending on the side of being defensive here as it seems like `skipLocsBefore` (and therefore `shouldSkip`) can also empty `LocsToResolve`, and we don't check for `isDone()` after calling them. No tests break if I remove these checks though. Ideally `shouldSkip` would propagate back either a "skip children" or "stop" walk action to return, but that's a little awkward to do with the templating and pre/post types. 

(as an aside, I'm starting to reconsider whether it's worth having different pre/post types for the walk actions, it is nice to statically reject returning `Action::SkipChildren()` in a post-walk, but it also makes ternaries and lambdas more awkward because Clang is fairly limited in its type inference)

bnbarham replied:
I see, I missed those ones. I think I'd personally prefer just having a `isDone` check in `tryResolve` itself rather than having to make sure we call `isDone` before calling it (or we crash because we use `.back()`). Ie. the pre/post would just stop at the next opportunity.

Anyway, that's more of a fundamental change, so I'm happy to keep it the way it is for now.

> (as an aside, I'm starting to reconsider whether it's worth having different pre/post types for the walk actions, it is nice to statically reject returning Action::SkipChildren() in a post-walk, but it also makes ternaries and lambdas more awkward because Clang is fairly limited in its type inference)

What's an example here? I don't have a strong preference either way. Given how many of these you've changed you probably have the best idea of what makes sense.

hamishknight replied:
(sorry for the delayed response)

> I see, I missed those ones. I think I'd personally prefer just having a `isDone` check in `tryResolve` itself rather than having to make sure we call `isDone` before calling it (or we crash because we use `.back()`). Ie. the pre/post would just stop at the next opportunity.
> 
> Anyway, that's more of a fundamental change, so I'm happy to keep it the way it is for now.

That makes sense to me, though we'll need to figure out a nice way to return it, as currently the bool return means "was something resolved", though it seems like most callers aren't interested in the result. I think for now I'll land this patch as-is, and can look into that as a follow-up.

> What's an example here? I don't have a strong preference either way. Given how many of these you've changed you probably have the best idea of what makes sense.

So e.g `b ? Action::SkipChildren() : Action::Continue()` currently doesn't work as Clang expects both branches to have the exact same type, and isn't able to look at the contextual type and have each branch do the conversion first. This was the main reason I added the conditional variants (though I kinda prefer them over a ternary anyways).

The lambdas I ran into when attempting to replace TypeWalker's `Action` with the ASTWalker action types, Clang expects each `return` to share the exact same type, so if you have e.g:

```c++
foo([&](bool b) {
  if (b)
    return Action::SkipChildren();
  return Action::Continue();
});
```

Clang doesn't like that, and you have to provide the explicit return type, e.g:

```c++
foo([&](bool b) -> PreWalkAction {
  if (b)
    return Action::SkipChildren();
  return Action::Continue();
});
```

which ehh isn't great IMO. I think the tradeoff of losing the compile-time rejection of `SkipChildren` for post-walking while gaining more convenience might be worth it.

bnbarham replied:
I like the `*If` over there ternary as well and I wouldn't find adding `-> PreWalkAction` that bad 🤷‍♂️. I'm fine either way though. If you do go the same-type route, please add a comment on `SkipChildren` that says it acts as `Continue` in the post case 😄.

hamishknight replied:
Hmm actually `Action::Stop()` would still need an intermediate type, which would have the same awkwardness (unless we erased the difference between "action" and "result" walker types, but that doesn't seem great). Might be worth sticking with the current design after all

bnbarham commented on lib/IDE/SwiftSourceDocInfo.cpp:
```diff
@@ -261,28 +263,17 @@ ASTWalker::PreWalkAction NameMatcher::walkToDeclPre(Decl *D) {
-  return Action::SkipChildrenIf(isDone());
-}
-
-ASTWalker::PostWalkAction NameMatcher::walkToDeclPost(Decl *D) {
```
Nice 👍

hamishknight committed:
[AST] Ensure the ASTWalker always stops

Fix up a few cases where we weren't propagating
the stop action correctly, sprinkling in some
`LLVM_NODISCARD` to help avoid the issue in
the future.

hamishknight committed:
Use `Action::Stop` in more places

This used to be something we couldn't express in
certain walker methods. Update various places
where we can now use it.

hamishknight commented:
@swift-ci please test
